[RISCV] Compress unrelaxable lui when RVC attribute is present

nemanjai opened the pull request:
If the object being linked allows compressed instructions, relaxation is on and we are unable to relax the HI20 relocation, emit c.lui instead of lui if the immediate fits.

Commit:
[RISCV] Compress unrelaxable lui when RVC attribute is present

If the object being linked allows compressed instructions,
relaxation is on and we are unable to relax the HI20
relocation, emit c.lui instead of lui if the immediate
fits.

Commit:
Run clang-format

topperc commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -650,8 +650,36 @@ static void relaxHi20Lo12(const InputSection &sec, size_t i, uint64_t loc,
   if (!gp)
     return;
 
+  bool noRelaxation = false;
```
Do we need this variable? Can we use `!isInt<12>(r.sym->getVA(r.addend) - gp->getVA())` in the one place we use the variabe?

nemanjai replied:
Sure. This was added in anticipation of another reason we might decide not to relax that is added in  [this PR](https://github.com/llvm/llvm-project/pull/73624)

nemanjai commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -650,8 +650,36 @@ static void relaxHi20Lo12(const InputSection &sec, size_t i, uint64_t loc,
+      return;
+    if (imm == 0)
+      newInsn &= 0xDFFF;
+    uint16_t imm17 = extractBits(imm, 17, 17) << 12;
```
Oh, I see. That does indeed seem wrong. It should be
```
uint16_t imm17 = extractBits(imm, 5, 5) << 12;
uint16_t imm16_12 = extractBits(imm, 4, 0) << 2;
```

Commit:
Address review comments

- Extract the correct bits from the immediate
- Add a test for an offset that fits in 18 bits, but the upper is not 6 bits

Commit:
Remove single-use bool and check eflags per-file

nemanjai commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -650,8 +650,32 @@ static void relaxHi20Lo12(const InputSection &sec, size_t i, uint64_t loc,
+    uint16_t imm16_12 = extractBits(imm, 4, 0) << 2;
+    remove = 2;
+    sec.relaxAux->relocTypes[i] = R_RISCV_RVC_LUI;
+    sec.relaxAux->writes.push_back(newInsn | imm17 | imm16_12);
```
The final instruction is indeed written in `RISCV::relocate()`. While we technically don't need to add the immediate here (i.e. we could just write a zero), I figured there is no harm in putting it in there - I don't know if there is any code path where this is needed.
